source/h2_connection.c

```diff
@@ -845,10 +846,58 @@ int s_decoder_on_push_promise(uint32_t stream_id, uint32_t promised_stream_id, v
     return AWS_OP_SUCCESS;
 }
 
-int s_decoder_on_data(uint32_t stream_id, struct aws_byte_cursor data, void *userdata) {
+int s_decoder_on_data_begin(uint32_t stream_id, uint32_t payload_len, void *userdata) {
     struct aws_h2_connection *connection = userdata;
 
-    /* #TODO Send a window update frame back. */
+    /* A receiver that receives a flow-controlled frame MUST always account for its contribution against the connection
+     * flow-control window, unless the receiver treats this as a connection error */
+    if (connection->thread_data.window_size_self < payload_len) {
+        /* TODO: Figure out what kind error this is, I failed to find it from spec, and it is possible to happen */
+        return aws_raise_error(AWS_ERROR_INVALID_STATE);
```

Comment:
A receiver MAY respond with a stream error (Section 5.4.2) or connection error (Section 5.4.1) of type FLOW_CONTROL_ERROR if it is unable to accept a frame.